types/openlayers/openlayers-tests.ts

```diff
@@ -573,6 +577,11 @@ const vectorLayer: ol.layer.Vector = new ol.layer.Vector({
     zIndex: -1
 });
 
+const nullOrUndefinedStyleFunction = (feature: (ol.Feature|ol.render.Feature), resolution: number): (null|undefined) => {
+    return resolution < 10 ? null : undefined;
+};
+vectorLayer.setStyle(nullOrUndefinedStyleFunction);
```

Comment:
If this is the reason why you submitted this PR, I think there is a confusion: the `nullOrUndefinedStyleFunction` function may return `null` or `undefined` but it isn't `null` itself: the value passed to `setStyle` isn't `null`. However the type of the function is not valid to be passed to `setStyle`. A `StyleFunction` is defined as `(feature: (ol.Feature | ol.render.Feature), resolution: number) => (ol.style.Style | ol.style.Style[] | null)`. Your returned type is wrong.

Reply:
The reason why I created the PR is that an `ol.StyleFunction` can return a `null` style. That is part of the OpenLayers renderer design and means "do not draw this feature".

See https://github.com/openlayers/openlayers/blob/07cb678271d725504762b9fa12f59a99c660dc0c/src/ol/renderer/canvas/VectorLayer.js#L347-L354

While fixing the Typescript definition file I searched for other incorrect definitions and fixed them:
- `setStyle` methods where not accepting `null` and `undefined` styles;
-  `ol.FeatureStyleFunction` can also return `null`.

I clarified the tests. Indeed, `undefined` is not a correct returned value for a style function.

Reply:
The code in your link does not say a `StyleFunction` can return `null`:
```
const styleFunction = feature.getStyleFunction() || vectorLayer.getStyleFunction();
if (styleFunction) {
  styles = styleFunction(feature, resolution);
}
```
Here it's the `styleFunction` variable that can be `null` or `undefined`. It means the `getStyleFunction()` method can returns `null`. It isn't the same as saying a `StyleFunction` can return `null`. 

The official definition for `StyleFunction`: https://github.com/openlayers/openlayers/blob/v4.6.5/src/ol/typedefs.js#L293. `null` is not allowed.

And in this case the definition file is not correct: none of the `getStyleFunction` methods allows `undefined` to be returned (unlike the documentation).
* http://openlayers.org/en/v4.6.5/apidoc/ol.Feature.html#getStyleFunction
* http://openlayers.org/en/v4.6.5/apidoc/ol.layer.Vector.html#getStyleFunction
* http://openlayers.org/en/v4.6.5/apidoc/ol.source.ImageVector.html#getStyleFunction
So if something has to be changed in the definitions, it's these `getStyleFunction` methods.

Reply:
Hi @osechet,

OpenLayers types have been written for the Google Closure Compiler. Here are the [rules governing nullability](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#nullable-type). As you see, an Object (instance of a type, function) is by default `nullable`.

[In the code I pointed](https://github.com/openlayers/openlayers/blob/07cb678271d725504762b9fa12f59a99c660dc0c/src/ol/renderer/canvas/VectorLayer.js#L350-L354), just below the code you quoted, the result of the style function is tested for nullability:
```js
      if (styles) {
        const dirty = this.renderFeature(
          feature, resolution, pixelRatio, styles, replayGroup);
        this.dirty_ = this.dirty_ || dirty;
}
```
This means: "if the `styles` returned by the style function is null, skip rendering the feature".

Regarding the definition of the style function:
```js
/**
 * @typedef {function(this: ol.Feature, number):  (ol.style.Style|Array.<ol.style.Style>)}
 */
```
`ol.style.Style` is nullable in Google Closure Compiler types; the equivalent typescript definition should explicitly include `null` which is the purpose of this PR. Note that the rules are different for `undefined` where one should be explicit.

Regarding the `getStyleFunction`s in OpenLayers, they are correct because `undefined` is an explicitly possible value. See https://github.com/openlayers/openlayers/blob/v4.6.5/src/ol/layer/vector.js#L199-L203.

Reply:
> In the code I pointed, just below the code you quoted, the result of the style function is tested for nullability

My mistake, I just checked the line you pointed to, not the following lines.

I didn't know about the Google Closure Compiler rules. Good to know.

> Regarding the getStyleFunctions in OpenLayers, they are correct because undefined is an explicitly possible value. See https://github.com/openlayers/openlayers/blob/v4.6.5/src/ol/layer/vector.js#L199-L203.

The documentation says they may return `undefined` but this type definition does not allow it, i.e.:
* https://github.com/DefinitelyTyped/DefinitelyTyped/blob/26bdec344feb14e749760a80e5ec4cac01d90493/types/openlayers/index.d.ts#L1719
* https://github.com/DefinitelyTyped/DefinitelyTyped/blob/26bdec344feb14e749760a80e5ec4cac01d90493/types/openlayers/index.d.ts#L6416
* https://github.com/DefinitelyTyped/DefinitelyTyped/blob/26bdec344feb14e749760a80e5ec4cac01d90493/types/openlayers/index.d.ts#L8476

Reply:
OK, done.